tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs

```diff
@@ -0,0 +1,301 @@
+        {
+            //Arrange
+            string result = string.Empty;
+
+            IConfidentialClientApplication cca = CreateCCAWithProxy(s_baseURL);
+
+            MsalServiceException ex = await AssertException.TaskThrowsAsync<MsalServiceException>(async () =>
+            {
+                await cca
+                .AcquireTokenForClient(s_msi_scopes)
+                .WithManagedIdentity()
+                .ExecuteAsync().ConfigureAwait(false);
+            }).ConfigureAwait(false);
+
+            Assert.IsNotNull(ex);
+        }
+
+        [DataTestMethod]
+        [DataRow(MsiAzureResource.WebApp, "", DisplayName = "System Identity Web App")]
+        [DataRow(MsiAzureResource.Function, "", DisplayName = "System Identity Function App")]
+        [DataRow(MsiAzureResource.VM, "", DisplayName = "System Identity Virtual Machine")]
+        [DataRow(MsiAzureResource.WebApp, UserAssignedClientID, DisplayName = "User Identity Web App")]
+        [DataRow(MsiAzureResource.Function, UserAssignedClientID, DisplayName = "User Identity Function App")]
+        [DataRow(MsiAzureResource.VM, UserAssignedClientID, DisplayName = "User Identity Virtual Machine")]
+        [DataRow(MsiAzureResource.WebApp, UamiResourceId, DisplayName = "ResourceID Web App")]
+        [DataRow(MsiAzureResource.Function, UamiResourceId, DisplayName = "ResourceID Function App")]
+        [DataRow(MsiAzureResource.VM, UamiResourceId, DisplayName = "ResourceID Virtual Machine")]
+        public async Task AcquireMSITokenAsync(MsiAzureResource azureResource, string userIdentity)
+        {
+            //Arrange
+            using (new EnvVariableContext())
+            {
+                //Get the Environment Variables
+                Dictionary<string, string> envVariables = 
+                    await GetEnvironmentVariablesAsync(azureResource).ConfigureAwait(false);
+
+                //Set the Environment Variables
+                SetEnvironmentVariables(envVariables);
+
+                //form the http proxy URI 
+                string uri = s_baseURL + $"MSIToken?" +
+                    $"azureresource={azureResource}&uri=";
+
+                //Create CCA with Proxy
+                IConfidentialClientApplication cca = CreateCCAWithProxy(uri);
+
+                //Act
+                AuthenticationResult result = await cca
+                        .AcquireTokenForClient(s_msi_scopes)
+                        .WithManagedIdentity(userIdentity)
+                        .ExecuteAsync().ConfigureAwait(false);
+            
+                //Assert
+                //1. Access token
+                Assert.IsNotNull(result.AccessToken);
+
+                //2. First token response is from the MSI Endpoint
+                Assert.AreEqual(TokenSource.IdentityProvider, result.AuthenticationResultMetadata.TokenSource);
+
+                //3. Validate the ExpiresOn falls within a 24 hour range from now
```

Comment:
Is this true? Don't they cache some tokens as well?

Reply:
No there is no caching on their side.